### PR TITLE
Create `BaseSynthesizer`

### LINKED ISSUE #1012: Create BaseSynthesizer class
### Problem Description

_As a user, it would be helpful if the naming of everything in `SDV` was consistent and clear._

We want to rename all of the current models to synthesizers. This is the first issue in that process.

### Expected behavior

- Create a `single_table` module in sdv
- In this module create a `BaseSynthesizer` class
- Add an `__init__` to this class that sets all of the required attributes for all synthesizers. These include:
    -  metadata (SingleTableMetadata)
    - enforce_min_max_values (bool)
    - enforce_rounding (bool)
- Add a `get_metadata()` method
    - This returns the `SingleTableMetadata` object
- Add a `get_parameters` method
    - This returns a dictionary of the parameters in the synthesizer minus the metadata. eg.
```python
model.get_parameters()
{
  'enforce_min_max_values': True,
  'enforce_rounding': False,
  'numerical_column_distributions': {
    'age': 'beta',
    'height': 'gaussian'
  }
}
````

### Additional context

- We may want to create the `DataProcessor` in the init as well, although that will be a private attribute

### PR DESCRIPTION
**Pull request by pvk-developer**
Resolves #1012

### CODE REVIEW

File: tests/unit/single_table/test_base.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    def test___init__(self, mock_data_processor):
+        """Test instantiating with default values."""
+        # Setup
+        metadata = Mock()
+
+        # Run
+        instance = BaseSynthesizer(metadata)
+
+        # Assert
+        assert instance.enforce_min_max_values
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:18]: I'd assert that it `is True`, just to make sure no other values are being passed. Same in the line below.

**Review by fealho — APPROVED** [review]
Looks good, just left one comment.

I think you did a good job with the tests as well, the docstrings are concise and explain the tests well.

File: tests/unit/single_table/test_base.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+        # Assert
+        assert instance.enforce_min_max_values is False
+        assert instance.enforce_rounding is False
+        assert instance._data_processor == mock_data_processor.return_value
+        mock_data_processor.assert_called_once_with(metadata)
+
+    @patch('sdv.single_table.base.DataProcessor')
+    def test_get_parameters(self, mock_data_processor):
+        """Test that returns every ``init`` parameter without the ``metadata``."""
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:40]: can we say ``Test that it returns``? The reads a little off

File: tests/unit/single_table/test_base.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        # Run
+        parameters = instance.get_parameters()
+
+        # Assert
+        assert 'metadata' not in parameters
+        assert parameters == {'enforce_min_max_values': False, 'enforce_rounding': False}
+
+    @patch('sdv.single_table.base.DataProcessor')
+    def test_get_metadata(self, mock_data_processor):
+        """Test that returns the ``metadata`` object."""
```
Comment (amontanez24) [inline tests/unit/single_table/test_base.py:54]: same

**Review by amontanez24 — APPROVED** [review]
One minor comment about the test docstrings but besides that looks great! :shipit:

**Pull request merged**